packages/utils/src/synergy.ts

```diff
@@ -0,0 +1,701 @@
+import { Ad4mClient, LinkQuery } from "@coasys/ad4m";
+import { getAd4mClient } from "@coasys/ad4m-connect/utils";
+import {
+  Conversation,
+  ConversationSubgroup,
+  Message,
+  Post,
+  Relationship,
+  SubjectRepository,
+  Topic,
+} from "@coasys/flux-api";
+import { languages } from "@coasys/flux-constants";
+import OpenAI from "openai";
+
+async function removeEmbedding(perspective, itemId) {
+  const embeddingLink = await perspective.get(
+    new LinkQuery({ source: itemId, predicate: "ad4m://embedding" })
+  );
+  if (embeddingLink[0]) return perspective.remove(embeddingLink[0]);
+}
+
+async function removeTopics(perspective, itemId) {
+  const relationships = (await findRelationships(perspective, itemId)) as any;
+  return Promise.all(
+    relationships.map(async (r) => {
+      const itemRelationshipLink = await perspective.get(
+        new LinkQuery({ source: itemId, target: r.id })
+      );
+      const relationshipTagLink = await perspective.get(
+        new LinkQuery({ source: r.id, predicate: "flux://has_tag" })
+      );
+      const linksToRemove = [] as any[];
+      if (itemRelationshipLink[0]) linksToRemove.push(itemRelationshipLink[0]);
+      if (relationshipTagLink[0]) linksToRemove.push(relationshipTagLink[0]);
+      return await perspective.removeLinks(linksToRemove);
+    })
+  );
+}
+
+async function removeProcessedData(perspective, itemId) {
+  return await Promise.all([
+    removeEmbedding(perspective, itemId),
+    removeTopics(perspective, itemId),
+  ]);
+}
+
+export async function generateEmbedding(text: string) {
+  const client = await getAd4mClient();
+  const embedding = await client.ai.embed("berd", text);
+  return embedding;
+}
+
+async function saveEmbedding(perspective, itemId, embedding) {
+  const { EMBEDDING_VECTOR_LANGUAGE } = languages;
+  const embeddingExpression = await perspective.createExpression(
+    { model: "berd", data: embedding },
+    EMBEDDING_VECTOR_LANGUAGE
+  );
+  return await perspective.add({
+    source: itemId,
+    predicate: "ad4m://embedding",
+    target: embeddingExpression,
+  });
+}
+
+async function getConversationData(perspective, conversationRepo) {
+  // check for previous conversations in the channel
+  const conversations = (await conversationRepo.getAllData()) as any;
+  let conversation;
+  let latestSubgroups = [];
+  let latestSubgroupItems = [] as any[];
+  if (conversations.length) {
+    // gather up lastest conversation data
+    const latestConversation = conversations[conversations.length - 1];
+    const subgroupRepo = await new SubjectRepository(ConversationSubgroup, {
```

Comment:
We decided to phase out `SubjectRepository` and have Flux only use `SubjectEntity` from ADAM client. I think we should change this here and if there are problems with what `SubjectEntity` provides, we should fix those or at least take note..